python/ray/data/_internal/stats.py

```diff
@@ -548,12 +558,66 @@ def update_dataset(self, dataset_tag: str, state: Dict[str, Any]):
             state_string = op_state.get("state", DatasetState.UNKNOWN.name)
             state_enum = DatasetState.from_string(state_string)
             self.data_operator_state.set(state_enum.value, operator_tags)
+            self.update_dataset_metadata_operator_state(
+                dataset_tag, operator, state_string
+            )
 
     def get_datasets(self, job_id: Optional[str] = None):
         if not job_id:
             return self.datasets
         return {k: v for k, v in self.datasets.items() if v["job_id"] == job_id}
 
+    def update_dataset_metadata_state(self, dataset_id: str, new_state: str):
+        if dataset_id not in self.dataset_metadatas:
+            return
+        update_time = time.time()
+        dataset_metadata = self.dataset_metadatas[dataset_id]
+        if dataset_metadata.state == new_state:
+            return
+        updated_dataset_metadata = copy.deepcopy(dataset_metadata)
+        updated_dataset_metadata.state = new_state
+        if new_state == DatasetState.RUNNING.name:
+            updated_dataset_metadata.execution_start_time = update_time
+        elif new_state in (DatasetState.FINISHED.name, DatasetState.FAILED.name):
+            updated_dataset_metadata.execution_end_time = update_time
+            # Update metadata of running operators
+            for operator in updated_dataset_metadata.topology.operators:
+                if operator.state == DatasetState.RUNNING.name:
+                    operator.state = new_state
+                    operator.execution_end_time = update_time
+
+        self.dataset_metadatas[dataset_id] = updated_dataset_metadata
+        self._metadata_exporter.export_dataset_metadata(updated_dataset_metadata)
+
+    def update_dataset_metadata_operator_state(
+        self, dataset_id: str, operator_id: str, new_state: str
+    ):
+        if dataset_id not in self.dataset_metadatas:
+            return
+        update_time = time.time()
+        dataset_metadata = self.dataset_metadatas[dataset_id]
+        updated_dataset_metadata = copy.deepcopy(dataset_metadata)
```

Comment:
This will be making N copies, where N is the number of operators in the dataset.

Instead let's make just 1 copy and update all states at once